samples: drivers: video: add capture to lvgl sample

CharlesDias opened the pull request:
Add sample application to capture an image frame from a camera and send it for display on the LCD.

This PR depends on:

- https://github.com/zephyrproject-rtos/zephyr/pull/71962
- https://github.com/zephyrproject-rtos/zephyr/pull/71462
- https://github.com/zephyrproject-rtos/hal_stm32/pull/204
- https://github.com/zephyrproject-rtos/hal_stm32/pull/195
- https://github.com/zephyrproject-rtos/zephyr/pull/76920
- https://github.com/zephyrproject-rtos/zephyr/pull/76921

CharlesDias committed:
samples: drivers: video: add capture to lvgl sample

Add sample application to capture an image frame from a camera
and send it for display on LCD via the LVGL library.

Signed-off-by: Charles Dias <charlesdias.cd@outlook.com>

CharlesDias commented:
Dear @kartben  and @nashif.

The goal is to use this example to validate the STM32 DMCI driver (PR https://github.com/zephyrproject-rtos/zephyr/pull/71462).

This sample has been tested on the mini_stm32h743 target with the OV2640 camera model. It depends on the PR https://github.com/zephyrproject-rtos/zephyr/pull/71462.

Should I move it to board-specific samples folder?

Thanks!

https://github.com/zephyrproject-rtos/zephyr/assets/27859307/7ef40ca5-7914-4f91-8a40-c53aff378782

CharlesDias commented:
Fix compliance check errors. 

The error "Found references to undefined Kconfig symbols" happened because this PR depend on the PR https://github.com/zephyrproject-rtos/zephyr/pull/71462

CharlesDias commented:
Fix image alignment

https://github.com/zephyrproject-rtos/zephyr/assets/27859307/53fbbd36-9375-454e-972f-80eab7d2dbeb

CharlesDias commented:
Add the clock rate control property. This property depend on https://github.com/zephyrproject-rtos/zephyr/pull/71962

CharlesDias commented:
Remove shell device, increase main stack size, and change the video buffer array from 1 to 2.

erwango commented on samples/subsys/video/capture_to_lcd/boards/mini_stm32h743.conf:
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+CONFIG_CLOCK_STM32_MCO1_SRC_HSI48=y
+CONFIG_CLOCK_STM32_MCO1_DIV=4
```
Is MCO requested for the sample to work ?

CharlesDias replied:
Yes. This board doesn't have a crystal oscillator for the camera module. The XCLK is provided by pin PA8.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/27859307/e96115d8-4d0d-413d-840e-d32111411b8f)

erwango commented on samples/subsys/video/capture_to_lcd/boards/mini_stm32h743.overlay:
```diff
@@ -0,0 +1,170 @@
+	/* DCMI_HSYNC */
+
+	/omit-if-no-ref/ dcmi_hsync_pa4: dcmi_hsync_pa4 {
+		pinmux = <STM32_PINMUX('A', 4, AF13)>;
```
Rather than defining DCMI pins here, please provide definitions in hal_stm32.
See https://github.com/zephyrproject-rtos/hal_stm32/pull/197 for an example on how to do that.

CharlesDias replied:
Hi, @erwango. The PRs have been opened:

- https://github.com/zephyrproject-rtos/hal_stm32/pull/204
- https://github.com/zephyrproject-rtos/hal_stm32/pull/205

josuah commented:
As I understand it, with the LVGL addition, this makes it closer to a *demo* (showing most of the capabilities of a system) than a *sample code* (minimal code to implement a single feature).

Is LVGL required for the video to display flow to work?
Maybe splitting/moving it to a separate sample inside `samples/module/lvgl` since it requires the LVGL module?

josuah commented on samples/subsys/video/capture_to_lcd/src/main.c:
```diff
@@ -0,0 +1,223 @@
+	}
+
+	/* Default to software video pattern generator */
+	video = device_get_binding(VIDEO_DEV_SW);
```
This is present in the original [video sample]() but the preferred API changed since then it seems?
https://github.com/zephyrproject-rtos/zephyr/blob/55c8a1650235cbf26145160b149a7f6bce2954eb/doc/releases/migration-guide-3.6.rst#devicetree-labels
https://github.com/zephyrproject-rtos/zephyr/blob/55c8a1650235cbf26145160b149a7f6bce2954eb/doc/releases/release-notes-3.2.rst#devicetree

CharlesDias replied:
Considering that the goal here is to display the frames capture by a real camera, I can remove the code below:

```
	video = device_get_binding(VIDEO_DEV_SW);
	if (video == NULL) {
		LOG_ERR("Video device %s not found", VIDEO_DEV_SW);
		return 0;
	}
```

josuah replied:
As see [here](https://github.com/zephyrproject-rtos/zephyr/blob/8ae0934470c9cd2bc56f4f5c4308c8b3f59485e9/drivers/video/Kconfig.sw_generator#L7-L9), there is no [safeguard](https://github.com/zephyrproject-rtos/zephyr/blob/8ae0934470c9cd2bc56f4f5c4308c8b3f59485e9/drivers/video/Kconfig.ov2640#L9) to check that the SW generator is on the device tree.  This is understandable as it was not updated since 5 years ago, where API might have been different.

This looks like lacking something like a device-tree `chosen` block to make the example generic, and then using [`DT_CHOSEN()`](https://docs.zephyrproject.org/latest/build/dts/api/api.html#c.DT_CHOSEN) here?

```c
video = DT_CHOSEN(camera);
```

These are all just suggestions: I am by no way endorsed by Zephyr in any way. Just offering feedback. :)

CharlesDias replied:
I appreciate your feedback! I'll test this suggestion once the required PRs are merged. Thanks!

josuah replied:
This also needs another pull request (I will try to make one) to add a `#define DT_DRV_COMPAT` and YAML binding so that the video pattern generator can be used on a devicetree, so best to keep it as you did until then I think.

CharlesDias commented:
> As I understand it, with the LVGL addition, this makes it closer to a _demo_ (showing most of the capabilities of a system) than a _sample code_ (minimal code to implement a single feature).

Thanks for your comments!

The goal was to create a similar sample as [tcpserversink](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/subsys/video/tcpserversink), in which captured frames are sent over an Ethernet connection.

I can remove the splash screen code and other parts to keep only the code needed to display the images on the LCD. How does that sound?

josuah commented:
I read the example source some more, and it looks like [LVGL is actually needed]() for getting it to work: [`lv_img_set_src`](https://github.com/zephyrproject-rtos/zephyr/pull/71463/files#diff-f0b70f9fa11c1c9756b892d32e0287960e132792af3b4e9759989137773f5f3bR212).

But since you implemented the video API as well for your [DMCI driver](
https://github.com/zephyrproject-rtos/zephyr/pull/71462/files), it is tempting to just pass the buffers around from the camera to the display. With the buffer dequeued from the camera is [directly enqueued to the display](https://github.com/zephyrproject-rtos/zephyr/assets/55351021/c35435cc-e367-4000-aadf-29998f102954) (from [here]([original talk](https://static.linaro.org/connect/san19/presentations/san19-503.pdf)).

Maybe the LVGL code is still good to have on the [LVGL samples](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/modules/lvgl)? to show how a mixed source VLGL application can be made.

I will order these WeAct boards but will not have time to try in-depth this week~month I think.
Thank you!

CharlesDias commented:
> But since you implemented the video API as well for your [DMCI driver](https://github.com/zephyrproject-rtos/zephyr/pull/71462/files), it is tempting to just pass the buffers around from the sensor to the display. With the buffer dequeued from the sensor is [directly enqueued to the display](https://github.com/zephyrproject-rtos/zephyr/assets/55351021/c35435cc-e367-4000-aadf-29998f102954) (from [here](https://static.linaro.org/connect/san19/presentations/san19-503.pdf)).

I'll take a close look at "Video 4 Zephyr". Thanks for share!

josuah commented:
Silly me! Of course LVGL is required! We do not have a camera pixel format, resolution and timing matching exactly the display parameters!

There need to be a bit of correction/processing, and LVGL does it perfectly as we can see!

There can be other models for resizing images using hardware acceleration (good fit for the `<drivers/video.h>` API too), but as of today, the way to go is to use LVGL. I am not even a maintainer, but LGTM!

=> **Ignore all of my previous comments, sorry for the confusion!**

CharlesDias commented:
Remove pinctrl node from overlay.

josuah commented:
There are now two pull requests implementing something similar: a full chain from a camera to a display.
https://github.com/zephyrproject-rtos/zephyr/pull/72633 uses the display drivers API directly without LVGL.

This is different from this example, focusing on integration with LVGL. How about including `lvgl` in the sample name to cohabitate better with this other PR?

loicpoulain commented:
Yes agree, should probably be something like `capture_to_lvgl` sample

CharlesDias commented:
Thank you guys for the suggestions.

I realized that this board configuration was broken after the https://github.com/zephyrproject-rtos/zephyr/pull/70583 that migrate the ST7735R to MIPI DBI. I opened this https://github.com/zephyrproject-rtos/zephyr/pull/73242 to fix.

CharlesDias commented:
Rebase

CharlesDias commented:
Hi, @josuah and @loicpoulain,

I changed the project name to `capture_to_lvgl` and I made small changes to the sample code.

josuah commented:
Nice!

The CI seems unhappy about small details, I think that'd need to be fixed before merging or everyone would have these CI failure too.

For instance:
https://github.com/zephyrproject-rtos/zephyr/actions/runs/9338622002/job/25701971209?pr=71463#step:9:57
https://github.com/zephyrproject-rtos/zephyr/actions/runs/9338621972/job/25701962635?pr=71463#step:8:38

I always forget to check CI myself...

CharlesDias commented:
Hi, @josuah,

Do you have any idea how to fix them? For example, [this one](https://github.com/zephyrproject-rtos/zephyr/actions/runs/9338622002/job/25701971209?pr=71463#step:9:57) complaint about the `undefined label: 'mini_stm32h743'`, but this is the board name.

Thanks for any help!

josuah commented:
It looks like you found all the missing CI fixes! I learned a bit about reStructuredText from you commits...

CharlesDias commented:
Rebase.

josuah commented on samples/subsys/video/capture_to_lvgl/src/main.c:
```diff
@@ -0,0 +1,215 @@
+int main(void)
+{
+	struct video_buffer *buffers[2], *vbuf;
+	const uint16_t WIDTH_VIDEO = CONFIG_VIDEO_WIDTH;
+	const uint16_t HEIGHT_VIDEO = CONFIG_VIDEO_HEIGHT;
```
Is this idiomatic of how to do things in LVGL?
Would it work to use `CONFIG_VIDEO_WIDTH` everywhere instead of `VIDEO_WDITH`?

CharlesDias replied:
You are totally right! Neither `WIDTH_VIDEO` nor `HEIGHT_VIDEO` is necessary. I'll update the sample.

Thanks, @josuah!

josuah reviewed (commented):
Hello again!
I confirm that this works! Same results as in your video: some few frames of noise then the logo from LVGL, then the video frames from the image sensor.

CharlesDias commented:
Hello, @josuah. 

Do you have any other suggestion? Thanks!

josuah commented on samples/subsys/video/capture_to_lvgl/src/main.c:
```diff
@@ -0,0 +1,213 @@
+		return 0;
+	}
+
+	/* Get format */
+	if (video_get_format(video_dev, VIDEO_EP_OUT, &fmt)) {
+		LOG_ERR("Unable to retrieve video format");
+		return 0;
+	}
```
Now that every field is set (see below), it might be possible to skip this?
Though, it might be convenient to debug.

CharlesDias replied:
Yes! I'll remove it!

CharlesDias replied:
Done

josuah reviewed (changes requested):
`fmt.pixelformat` change proposed, the others are plain feedback.

Sorry for not spotting that earlier!

kartben commented on samples/subsys/video/capture_to_lvgl/src/main.c:
```diff
@@ -0,0 +1,213 @@
+{
+	lv_obj_t *label = lv_label_create(lv_scr_act());
+
+	lv_label_set_text(label, "LVGL on Zephyr!");
```
```suggestion
	lv_label_set_text(label, "Video capture sample");
```

CharlesDias replied:
The splash sccreen was removed.

kartben commented on samples/subsys/video/capture_to_lvgl/src/main.c:
```diff
@@ -0,0 +1,213 @@
+	while (caps.format_caps[i].pixelformat) {
+		const struct video_format_cap *fcap = &caps.format_caps[i];
+		/* fourcc to string */
+		printk("  %c%c%c%c width [%u; %u; %u] height [%u; %u; %u]\n",
```
please log with appropriate level instead of using printk

CharlesDias replied:
Hi @kartben, thanks for your review!

Should I declare the LOG_MODULE_REGISTER macro to this code?

CharlesDias replied:
Changed to LOG_INF

kartben commented on samples/subsys/video/capture_to_lvgl/src/main.c:
```diff
@@ -0,0 +1,213 @@
+void display_splash_screen(void)
+{
+	draw_label();
+	draw_lvgl_logo();
```
is this really needed? Unless I am missing something I don't see the added value of this image (but it certainly adds to the binary image size :))

In fact, I am really questioning the justification of having a splash screen in the first place, and I would jump straight into the actual video sample.

kartben replied:
now that I think about it some more, I realize that maybe your intention was to show "a bit more" of LVGL in your sample than just video stuff. In this case, what I think would be cool, and likely pretty simple, would be to show how easy it is to add the ability to drag the video (of course this requires a touch screen) https://docs.lvgl.io/master/examples.html#make-an-object-draggable ... or something similar?

CharlesDias replied:
I agree to keep it simple. 

About the drag suggestion, can we leave this improvement for the future?

CharlesDias replied:
The splash screen was removed.

kartben commented on samples/subsys/video/capture_to_lvgl/README.rst:
```diff
@@ -0,0 +1,82 @@
+
+This sample requires a video capture device (e.g., a camera) and an LCD display.
+
+- :ref:`mini_stm32h743`
+- OV2640 camera module
+- LCD display
```
not sure what this list is supposed to mean?

CharlesDias replied:
The intention was to list the necessary devices, but I believe the above description is sufficient. I'll remove it.

CharlesDias replied:
Removed

kartben commented on samples/subsys/video/capture_to_lvgl/README.rst:
```diff
@@ -0,0 +1,82 @@
+Requirements
+************
+
+This sample requires a video capture device (e.g., a camera) and an LCD display.
```
doesn't have to be *LCD* specifically, right?

```suggestion
This sample requires a supported :ref:`video capture device <video_api>` (e.g., a camera) and a :ref:`display <display_api>`.
```

CharlesDias replied:
Changed

kartben commented on samples/subsys/video/capture_to_lvgl/README.rst:
```diff
@@ -0,0 +1,82 @@
+   :zephyr-app: samples/subsys/video/capture_to_lvgl/
+   :board: mini_stm32h743
+   :goals: build flash
+   :gen-args: -DCONFIG_BOOT_DELAY=5000
```
really needed?

CharlesDias replied:
Yes. Since the board uses USB CDC to print the log messages, some messages are missing when the delay is removed, but I can reduce it from 5000 to 2000.

CharlesDias replied:
Changed to 2000

kartben commented on samples/subsys/video/capture_to_lvgl/boards/mini_stm32h743.overlay:
```diff
@@ -0,0 +1,85 @@
+	status = "okay";
+	clock-frequency = <I2C_BITRATE_FAST>;
+
+	ov2640: ov2640@30 {
```
what is the form factor / connector of the camera you're using? It feels like it would maybe be more reusable if you would be defining it as a shield?

CharlesDias replied:
Please, see the image.

![image](https://github.com/user-attachments/assets/b56f731f-73d5-4763-b7f7-941e6f66488c)

josuah replied:
It looks like the connector of the ESP32-CAM modules made by various vendors.
https://timsblogplace.blogspot.com/2022/01/esp32-cam-rotating-camera.html

It looks like there are several cameras module proposed by WeAct Studio, that are all expected to fit into that connector. Some def-facto ESP32-CAM connector maybe?

I confirm that the FPC of OV7725 and OV2640 here both fit mechanically at least:
https://www.aliexpress.com/item/1005006632336183.html

CharlesDias replied:
You are right! This board also supports the OV5640.

@kartben and @josuah, should I implement it as a shield?

josuah replied:
> should I implement it as a shield?

If you have the patience to, I think this would be appreciated!

kartben commented on samples/subsys/video/capture_to_lvgl/prj.conf:
```diff
@@ -0,0 +1,20 @@
+CONFIG_LVGL=y
+CONFIG_LV_MEM_CUSTOM=y
+CONFIG_LV_USE_LOG=y
+CONFIG_LV_USE_LABEL=y
+CONFIG_LV_USE_BTN=y
+CONFIG_LV_USE_ARC=y
+CONFIG_LV_USE_IMG=y
```
given you are not using `LV_CONF_MINIMAL`, I don't think these are necessary as they already default to `y`.
That being said, it would be nice to use `LV_CONF_MINIMAL` and make sure that only the LVGL feature you need are turned on as this will significantly reduce the ROM usage.

CharlesDias replied:
Is this configuration below ok? I have to keep the `CONFIG_LV_USE_IMG=y`, otherwise I get a compile error regarding the `lv_img_create` function.

```
CONFIG_LVGL=y
CONFIG_LV_CONF_MINIMAL=y
CONFIG_LV_MEM_CUSTOM=y
CONFIG_LV_USE_IMG=y
CONFIG_LV_Z_MEM_POOL_SIZE=16384
```

kartben reviewed (changes requested):
need to find some time to actually test this, but some comments already -- thanks for the great work @CharlesDias!

kartben commented on samples/subsys/video/capture_to_lvgl/src/main.c:
```diff
@@ -0,0 +1,213 @@
+
+		frame_counter++;
+
+		/* Calculate FPS every second */
+		if (k_uptime_get_32() - last_time > 1000) {
+			fps = frame_counter;
+
+			/* Reset frame counter and timer */
+			frame_counter = 0;
+			last_time = k_uptime_get_32();
+		}
+
+		printk("\rFrame: %u, FPS: %u, Size: %u bytes, Timestamp: %u ms",
+			frame++, fps, vbuf->bytesused, vbuf->timestamp);
```
related to my comment on showing "more" lvgl stuff -- I would totally display the framerate on the screen... in fact, you can probably get rid of your custom code altogether and simply enable `CONF_LV_USE_PERF_MONITOR` :)

CharlesDias replied:
Do you think I should have removed this print message and enabled CONF_LV_USE_PERF_MONITOR?

![photo_5028563151020469497_y](https://github.com/user-attachments/assets/f79b49e3-fe99-480b-9522-ec9145e33b00)

kartben replied:
I think it makes a lot of sense to just rely on upstream LVGL functionality whenever possible, no? :)
Cool to see it in action!

CharlesDias replied:
Removed the previous log lines and enabled the CONF_LV_USE_PERF_MONITOR.

CharlesDias commented:
Hello, @josuah and @kartben. 

Thank you so much for your review! I'll check them in a couple of days.

CharlesDias commented on samples/subsys/video/capture_to_lvgl/src/main.c:
```diff
@@ -0,0 +1,213 @@
+	       fmt.width, fmt.height);
+
+	/* Set format */
+	fmt.width = CONFIG_VIDEO_WIDTH;
+	fmt.height = CONFIG_VIDEO_HEIGHT;
+	fmt.pitch = fmt.width * 2;
```
It's a good idea to set up the format also. Thanks!

CharlesDias replied:
Done

CharlesDias commented on samples/subsys/video/capture_to_lvgl/src/main.c:
```diff
@@ -0,0 +1,213 @@
+		i++;
+	}
+
+	/* Get default/native format */
+	if (video_get_format(video_dev, VIDEO_EP_OUT, &fmt)) {
+		LOG_ERR("Unable to retrieve video format");
+		return 0;
+	}
+
+	printk("- Default format: %c%c%c%c %ux%u\n", (char)fmt.pixelformat,
+	       (char)(fmt.pixelformat >> 8),
+	       (char)(fmt.pixelformat >> 16),
+	       (char)(fmt.pixelformat >> 24),
+	       fmt.width, fmt.height);
```
Yes! I'll remove it!

CharlesDias replied:
Done

CharlesDias commented:
Rebase.

CharlesDias commented:
Implement code review recommendations.

CharlesDias commented:
> /* Video flip image horizontally */
> 	int hflip = CONFIG_VIDEO_HFILP;
> 
> 	int ret = video_set_ctrl(video_dev, VIDEO_CID_HFLIP, &hflip);
> 	if (ret) {
> 		LOG_ERR("Unable to set video control (HFLIP) %d", ret);
> 		return 0;
> 	}
> 
> 	/* Video flip image vertically */
> 	int vflip = CONFIG_VIDEO_VFILP;
> 
> 	ret = video_set_ctrl(video_dev, VIDEO_CID_VFLIP, &vflip);
> 	if (ret) {
> 		LOG_ERR("Unable to set video control (VFLIP) %d", ret);
> 		return 0;
> 	}

Change to 

```
#ifdef CONFIG_VIDEO_HFILP
	/* Video flip image horizontally */
	int hflip = 1;

	if (video_set_ctrl(video_dev, VIDEO_CID_HFLIP, &hflip)) {
		LOG_ERR("Unable to set video control (HFLIP)");
		return 0;
	}
#endif

#ifdef CONFIG_VIDEO_VFILP
	/* Video flip image vertically */
	int vflip = 1;

	if (video_set_ctrl(video_dev, VIDEO_CID_VFLIP, &vflip)) {
		LOG_ERR("Unable to set video control (VFLIP)");
		return 0;
	}
#endif
```

josuah commented on samples/subsys/video/capture_to_lvgl/src/main.c:
```diff
@@ -0,0 +1,166 @@
+	/* Video flip image horizontally */
+	int hflip = 1;
+
+	if (video_set_ctrl(video_dev, VIDEO_CID_HFLIP, &hflip)) {
```
I think the `.set_ctrl()` values are passed literally when they are integers, so you can replace `&hflip` by `(void *)1`.

Here are examples of how they are handled:
https://github.com/zephyrproject-rtos/zephyr/blob/1c689fce182623932843536abdb2e4e5ba4e003d/drivers/video/ov2640.c#L935-L940

https://github.com/zephyrproject-rtos/zephyr/blob/1c689fce182623932843536abdb2e4e5ba4e003d/drivers/video/video_sw_generator.c#L254-L256

CharlesDias replied:
Done. Thanks!

josuah reviewed (approved):
I leave it to @kartben regarding the [shield](https://github.com/zephyrproject-rtos/zephyr/pull/71463/files#r1710246857).

Suggesting to remove the `&` to `video_set_ctrl()`.

Thank you for improving this example, getting more mature through review cycles, and very convenient for testing compatible image sensors!

[EDIT: I hit "approve" instead of "request changes", though these are the last things I could notice]

CharlesDias commented:
Rebase

josuah commented on samples/subsys/video/capture_to_lvgl/CMakeLists.txt:
Sorry about this: now that https://github.com/zephyrproject-rtos/zephyr/pull/73001 is merged, this might need to move to `samples/drivers/` as well...

josuah reviewed (changes requested):
Some very minor changes noted above (shield?, moving samples), but now all PR dependencies are merged! :tada:

CharlesDias commented:
Hi, @josuah and @kartben.

I opened the PR https://github.com/zephyrproject-rtos/zephyr/pull/77949 related to the shield suggestion. Could you take a look? Thanks1

CharlesDias commented:
> Some very minor changes noted above (shield?, moving samples), but now all PR dependencies are merged! 🎉

I have added one more PR dependency! hahaha

CharlesDias commented:
Move the sample to `samples/drivers/` and update the overlay file due the shield implemented by https://github.com/zephyrproject-rtos/zephyr/pull/77949

josuah commented on samples/drivers/video/capture_to_lvgl/boards/mini_stm32h743.conf:
```diff
@@ -0,0 +1,15 @@
+
+CONFIG_VIDEO_BUFFER_POOL_SZ_MAX=102400
+
+CONFIG_VIDEO_HFILP=y
+CONFIG_VIDEO_VFILP=y
```
`FILP` -> `FLIP`

CharlesDias replied:
Fixed and configuration moved to shield https://github.com/zephyrproject-rtos/zephyr/pull/77949.

josuah commented on samples/drivers/video/capture_to_lvgl/Kconfig:
```diff
@@ -0,0 +1,22 @@
+	int "Define the height of the video"
+	default 120
+
+config VIDEO_HFILP
+	bool "Horizontal flip"
+	default n
+
+config VIDEO_VFILP
+	bool "Vertical flip"
+	default n
```
`FILP` -> `FLIP`

CharlesDias replied:
Fixed. Thanks!

josuah commented on samples/drivers/video/capture_to_lvgl/src/main.c:
```diff
@@ -0,0 +1,162 @@
+		video_enqueue(video_dev, VIDEO_EP_OUT, buffers[i]);
+	}
+
+#ifdef CONFIG_VIDEO_HFILP
+	/* Video flip image horizontally */
+	if (video_set_ctrl(video_dev, VIDEO_CID_HFLIP, (void *)1)) {
+		LOG_ERR("Unable to set video control (HFLIP)");
+		return 0;
+	}
+#endif
+
+#ifdef CONFIG_VIDEO_VFILP
+	/* Video flip image vertically */
+	if (video_set_ctrl(video_dev, VIDEO_CID_VFLIP, (void *)1)) {
+		LOG_ERR("Unable to set video control (VFLIP)");
+		return 0;
+	}
+#endif
```
`FILP` -> `FLIP`

CharlesDias replied:
Fixed. Thanks!

josuah reviewed (changes requested):
Some tiny typo spotted.

Maybe this can be merged before the shield support: it is possible to run this with the pattern generator. Though some incorrect doc until the shield lands in. That depends on available patience!

CharlesDias commented:
> Maybe this can be merged before the shield support: it is possible to run this with the pattern generator. Though some incorrect doc until the shield lands in. That depends on available patience!

Hi, @josuah. 

To run with the pattern generator, I have to comment out the code below, otherwise I get a compiler error. Should I move these code below to the shield?

`capture_to_lvgl/boards/mini_stm32h743.conf`
```
CONFIG_VIDEO_HFILP=y
CONFIG_VIDEO_VFILP=y
```

and `capture_to_lvgl/boards/mini_stm32h743.overlay`
```
/ {
	chosen {
		zephyr,camera = &camera_interface;
	};
};
```

![photo_5134583246196157781_x](https://github.com/user-attachments/assets/3576e289-abda-4474-a7e4-00cbb41a4290)

CharlesDias commented:
Fixed typos and move out some configuration to shield https://github.com/zephyrproject-rtos/zephyr/pull/77949. This version works with "VIDEO_SW_GENERATOR" and with shield https://github.com/zephyrproject-rtos/zephyr/pull/77949. 

@josuah and @kartben, I think it is ready to go! This PR doesn't depend on the PR shield anymore.

josuah commented:
@CharlesDias This is a good idea. Then every shield can specify what they bring to the table an image sensor, a display, both...).
Some camera will be mounted upside down, others not.

CharlesDias commented:
The pipeline failed, but it didn't provide a reason!

![image](https://github.com/user-attachments/assets/ec1cf648-948e-4a9f-a4f8-1e86fbc199f9)

josuah commented:
It looks like an infrastructure-related issue. You can `git commit --amend` and force-push again to trigger a new CI cycle and it will hopefully be sorted out.

josuah commented:
Well, actually the shield is merged now! :)
So while moving the config to the shield is a good thing to do, you can now do this in this PR.
Doing so will also trigger a new CI cycle.

CharlesDias commented:
> Well, actually the shield is merged now! :)

Unfortunately, no! The shield PR is still open. I just removed it from the PR dependency list above.

- https://github.com/zephyrproject-rtos/zephyr/pull/77949

CharlesDias commented:
Only to force CI to run.

josuah reviewed (dismissed):
There is one last reference to something that is just about to be merged `:shield: weact_ministm32h7xx_ov2640`.
The rest all looks like being ready.

Some config like `CONFIG_VIDEO_HFLIP` or `DT_HAS_CHOSEN(zephyr_camera)` are not used by this example, but were still tested against a custom shield (being submitted), so are known to work, and can be used by anyone integrating their own shield in the future.

This currently works with the test pattern generator.

erwango commented on samples/drivers/video/capture_to_lvgl/boards/mini_stm32h743.conf:
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+CONFIG_CLOCK_STM32_MCO1_SRC_HSI48=y
+CONFIG_CLOCK_STM32_MCO1_DIV=4
```
Those symbols are now deprecated and MCO configuration should be done by device tree (see https://github.com/zephyrproject-rtos/zephyr/pull/76556)

CharlesDias replied:
Thanks, @erwango. Since this version works using the "VIDEO_SW_GENERATOR", I'm removing this configuration from here, and I'll add the new MCO configuration on `:shield: weact_ministm32h7xx_ov2640`, PR https://github.com/zephyrproject-rtos/zephyr/pull/77949.

erwango commented on samples/drivers/video/capture_to_lvgl/sample.yaml:
```diff
@@ -0,0 +1,10 @@
+  sample.video.capture_to_lvgl:
+    build_only: true
+    tags:
+      - video
+    platform_allow:
```
Non blocking: Shouldn't it be `integration_platform` if this is the reference platform you want this sample to be built with ?

CharlesDias replied:
Thanks. I added 
```
integration_platforms:
      - mini_stm32h743
```

erwango commented on samples/drivers/video/capture_to_lvgl/sample.yaml:
```diff
@@ -0,0 +1,10 @@
+      - video
+    platform_allow:
+      - mini_stm32h743
+    depends_on: video
```
If `video` not declared by board, CI won't build the test.

CharlesDias replied:
Sorry, I didn't get the problem. If the board does not declare the `video`, I believe that the `VIDEO_SW_GENERATOR` will be used.

CharlesDias replied:
I think I understood the issue! :) I added the `video` feature to `mini_stm32h743` board.

erwango commented on samples/drivers/video/capture_to_lvgl/boards/mini_stm32h743.overlay:
```diff
@@ -0,0 +1,31 @@
+	status = "okay";
+};
+
+&rcc {
```
Non blocking: If you're changing the clock for some specific reason, let's put it as a comment.

CharlesDias replied:
I'll remove this configuration because it's not necessary to use the "VIDEO_SW_GENERATOR".

CharlesDias commented:
Rebase.

CharlesDias commented:
Update the samples.yaml by adding the `integration_platforms`; removed the `pll` and `rcc` configuration from `mini_stm32h743.overlay`; changed the default video width and height to the 320x240 resolution because display on `native_sim/native/64` has the 320x240 size resolution.

josuah commented on samples/drivers/video/capture_to_lvgl/README.rst:
```diff
@@ -0,0 +1,74 @@
+Requirements
+************
+
+This sample requires a supported :ref:`video capture device <video_api>` (e.g., a camera)
+and a :ref:`display <display_api>`.
```
This implies that the `sw_generator` is a capture device... which is true to some extent!

josuah reviewed (dismissed):
This looks very minimal now. Easy to use as a starting point for graphics + video.

kartben commented on samples/drivers/video/capture_to_lvgl/README.rst:
```diff
@@ -0,0 +1,74 @@
+***********
+
+The application uses the :ref:`Video API <video_api>` to retrieve video frames from
+a video capture device, writes a frame count message to the console, and then sends
+the frame to an LCD display.
```
```suggestion
a video capture device, write a frame count message to the console, and then send
the frame to an LCD display.
```

CharlesDias replied:
Done.

CharlesDias commented on samples/drivers/video/capture_to_lvgl/README.rst:
```diff
@@ -0,0 +1,74 @@
+   :relevant-api: video_interface
+
+   This sample application demonstrates how to use the video API to retrieve video
+   frames from a capture device and display them on an LCD via the LVGL library.
```
Done. Thanks!

CharlesDias commented:
Implemented @kartben suggestions.
